File: LeetCode/42. 接雨水.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
  const n = height.length
  if (n === 0) return
  const leftMax = Array.from({ length: n }, (_, i) => i)
  leftMax[0] = height[0]
  for (let i = 1; i < n; i++) {
    leftMax[i] = Math.max(leftMax[i - 1], height[i])
  }
  const rightMax = Array.from({ length: n }, (_, i) => i)
  rightMax[n - 1] = height[n - 1]
  for (let i = n - 2; i >= 0; i--) {
    rightMax[i] = Math.max(rightMax[i + 1], height[i])
  }
  let answer = 0
  for (let i = 0; i < n; i++) {
    answer += Math.min(leftMax[i], rightMax[i]) - height[i]
  }
  return answer
}

File: LeetCode/1-100/42. 接雨水.go
func trap(height []int) int {
    n := len(height)
    ans := 0
    if n == 0 {
        return ans 
    }
    leftMax := make([]int, n)
    rightMax := make([]int, n)
    leftMax[0],rightMax[n-1] = height[0],height[n-1]
    for i:=1;i<n;i++ {
        leftMax[i] = max(leftMax[i - 1], height[i])
    }
    for i:=n-2;i>=0;i-- {
        rightMax[i] = max(rightMax[i + 1], height[i])
    }
    for i,v := range height {
        ans += min(leftMax[i],rightMax[i]) - v
    }
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
